src/Stratis.Bitcoin.Features.Consensus/StakeValidator.cs

```diff
@@ -45,13 +48,137 @@ static StakeValidator()
-            this.network = network;
             this.stakeChain = stakeChain;
             this.chain = chain;
             this.coinView = coinView;
             this.consensusOptions = network.Consensus.Option<PosConsensusOptions>();
         }
 
+        /// <summary>
+        /// Gets the last block in the chain that was generated using
+        /// PoS if <see cref="proofOfStake"/> is <c>true</c> or PoW if <see cref="proofOfStake"/> is <c>false</c>.
+        /// </summary>
+        /// <param name="stakeChain">Chain that is used for retrieving blocks.</param>
+        /// <param name="index">Block that we start from. Only blocks before that one will be checked.</param>
+        /// <param name="proofOfStake">Specifies what kind of block we are looking for: PoS or PoW.</param>
+        public static ChainedBlock GetLastBlockIndex(StakeChain stakeChain, ChainedBlock index, bool proofOfStake)
+        {
+            if (index == null)
+                throw new ArgumentNullException(nameof(index));
+
+            clogger.LogTrace("({0}:'{1}',{2}:{3})", nameof(index), index, nameof(proofOfStake), proofOfStake);
+
+            BlockStake blockStake = stakeChain.Get(index.HashBlock);
+
+            while ((index.Previous != null) && (blockStake.IsProofOfStake() != proofOfStake))
+            {
+                index = index.Previous;
+                blockStake = stakeChain.Get(index.HashBlock);
+            }
+
+            clogger.LogTrace("(-)':{0}'", index);
+            return index;
+        }
+
+        /// <summary>
+        /// Calculates the difficulty target for the next block.
+        /// </summary>
+        /// <param name="stakeChain">Database of stake related data for the current blockchain.</param>
+        /// <param name="chainedBlock">Block header for which to calculate the target difficulty.</param>
+        /// <param name="consensus">Consensus rules for the current network.</param>
+        /// <param name="proofOfStake"><c>true</c> for calculation of PoS difficulty target, <c>false</c> for calculation of PoW difficulty target.</param>
+        /// <returns>The difficulty target for the next block after <paramref name="chainedBlock"/>.</returns>
+        /// <remarks>
+        /// The calculation of the next target is based on the last target value and the block time (aka spacing) of <paramref name="chainedBlock"/>
+        /// (i.e. difference in time stamp of this block and its immediate predecessor). The target changes every block and it is adjusted
+        /// down (i.e. towards harder to reach, or more difficult) if the time to mine last block was lower than the target block time.
+        /// And it is adjusted up if it took longer than the target block time. The adjustments are done in a way the target is moving towards
+        /// the target spacing exponentially, so even a big change in the mining power on the network will be fixed by retargeting relatively quickly.
```

Comment:
Me neither- it's an old one, I just added xmldoc for it.

Reply:
I see so small fix, I think this is correct.
>The adjustments are done in a way where the target is moving towards the target-spacing exponentially

Reply:
target spacing is what normal ppl would call block time, or maybe rather the expected block time, or the target block time we want to achieve. if you want to write it as target-spacing, whatever

the idea is that once you are off your target, you change the difficulty, so that if no more and no less stake power is on the network next time, you would expect the next block time to be closer to your target - this forms a series which converges to the target and the speed of the converging is exponential